universal-application-tool-0.0.1/app/repository/ProgramRepository.java

```diff
@@ -68,61 +72,67 @@ public Program updateProgramSync(Program program) {
-                    .count()
-                == 1,
-            "Must be exactly one program with this name in the draft.");
-        versionRepository.get().updateQuestionVersions(newDraft);
-        transaction.commit();
-        return newDraft;
-      } catch (IllegalStateException e) {
-        transaction.rollback();
-        // We must end the transaction here since we are going to recurse and try again.
-        // We cannot have this transaction on the thread-local transaction stack when that
-        // happens.
-        transaction.end();
-        return createOrUpdateDraft(existingProgram);
-      } finally {
-        // This may come after a prior call to `transaction.end` in the event of a
-        // precondition failure - this is okay, since it a double-call to `end` on
-        // a particular transaction.  Only double calls to database.endTransaction
-        // must be avoided.
-        transaction.end();
-      }
+      return updateProgramSync(updatedDraft);
+    }
+
+    // Inside a question update, this will be a savepoint rather than a full transaction.  Otherwise
+    // it will be creating a new transaction.
+    Transaction transaction = database.beginTransaction(TxScope.required());
+    try {
+      // Program -> builder -> back to program in order to clear any metadata stored in the program
+      // (for example, version information).
+      Program newDraft = existingProgram.getProgramDefinition().toBuilder().build().toProgram();
+      newDraft = insertProgramSync(newDraft);
+      newDraft.addVersion(draftVersion).save();
+      draftVersion.refresh();
+      Preconditions.checkState(
+          draftVersion.getPrograms().contains(newDraft),
+          "Must have successfully added draft version.");
+      Preconditions.checkState(
+          draftVersion.getLifecycleStage().equals(LifecycleStage.DRAFT),
+          "Draft version must remain a draft throughout this transaction.");
+      // Ensure we didn't add a duplicate with other code running at the same time.
+      String programName = existingProgram.getProgramDefinition().adminName();
+      Preconditions.checkState(
+          draftVersion.getPrograms().stream()
+                  .map(Program::getProgramDefinition)
+                  .map(ProgramDefinition::adminName)
+                  .filter(programName::equals)
+                  .count()
+              == 1,
+          "Must be exactly one program with this name in the draft.");
+      versionRepository.get().updateQuestionVersions(newDraft);
+      transaction.commit();
+      return newDraft;
+    } catch (IllegalStateException e) {
+      transaction.rollback();
+      // We must end the transaction here since we are going to recurse and try again.
+      // We cannot have this transaction on the thread-local transaction stack when that
+      // happens.
+      transaction.end();
+      return createOrUpdateDraft(existingProgram);
```

Comment:
would this forever loop in some situations?

Reply:
seemingly so.... yeah that looked like it could be a problem.  I'll review this when I dig into how transactions work in our code.

Reply:
Just to be clear, this is a pre-existing issue, nothing I did changes it either way.